File: solutions/python/binary-search/find-the-duplicate-number.py
# LeetCode 287 — Find the Duplicate Number (Medium)
# Category: Binary Search · Approach: Binary Search on Value
# Time: O(n log n) | Space: O(1)
# Source: https://leetcode.com/problems/find-the-duplicate-number/

def find_duplicate(nums: list[int]) -> int:
    lo, hi = 1, len(nums) - 1
    while lo < hi:
        mid = (lo + hi) // 2
        count = sum(1 for x in nums if x <= mid)
        if count > mid:
            hi = mid
        else:
            lo = mid + 1
    return lo

File: solutions/typescript/binary-search/find-the-duplicate-number.ts
// LeetCode 287 — Find the Duplicate Number (Medium)
// Category: Binary Search · Approach: Binary Search on Value
// Time: O(n log n) | Space: O(1)
// Source: https://leetcode.com/problems/find-the-duplicate-number/

function findDuplicate(nums: number[]): number {
  let lo = 1, hi = nums.length - 1;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    let count = 0;
    for (const x of nums) if (x <= mid) count++;
    if (count > mid) hi = mid;
    else lo = mid + 1;
  }
  return lo;
}
